src/shared/bus-util.c

```diff
@@ -508,7 +510,15 @@ int bus_connect_transport_systemd(
                 switch (runtime_scope) {
 
                 case RUNTIME_SCOPE_USER:
-                        return bus_connect_user_systemd(ret_bus);
+                        r = bus_connect_user_systemd(ret_bus);
+                        /* We used to always fall back to the user session bus if we couldn't connect to the
+                         * private manager bus. To keep compat with existing code that was setting
+                         * DBUS_SESSION_BUS_ADDRESS without setting XDG_RUNTIME_DIR, connect to the user
+                         * session bus if DBUS_SESSION_BUS_ADDRESS is set and XDG_RUNTIME_DIR isn't. */
+                        if (r == -ENOMEDIUM && secure_getenv("DBUS_SESSION_BUS_ADDRESS"))
+                                r = sd_bus_default_user(ret_bus);
```

Comment:
Nah, I do not like such taping over kind of misconfiguration. Why not to just set XDG_RUNTIME_DIR?

Reply:
Because this used to work and we found code in the wild relying on this. This is about keeping backwards compat with existing code.

Reply:
Hm,,, but that's for `sd_bus_open_user()`, and it still supports the environment variable as documented. My question is that in this case, like systemctl, should we really fall back to use the function? How about to 'guess' the bus address for direct connection to the user service manager from `DBUS_SESSION_BUS_ADDRESS` ?

Reply:
> Hm,,, but that's for sd_bus_open_user(), and it still supports the environment variable as documented. My question is that in this case, like systemctl, should we really fall back to use the function?

That's a fair point

> How about to 'guess' the bus address for direct connection to the user service manager from DBUS_SESSION_BUS_ADDRESS ?

So if you trace through the code, we actually do use the user service manager from DBUS_SESSION_BUS_ADDRESS since sd_bus_default_user -> sd_bus_default_user -> sd_bus_open_user -> sd_bus_open_user_with_description -> bus_set_address_user which does read the bus address from the env variable: https://github.com/systemd/systemd/blob/b6b8527cd184085008585c1a1f9725eb97c342ef/src/libsystemd/sd-bus/sd-bus.c#L1373

With that said, yea its probably pretty brittle. Let me add an integration test to make sure there's no regression, then we can also figure out how to refactor and/or if to support DBUS_SESSION_BUS_ADDRESS in systemctl